src/controllers/generate.controller.ts

```diff
@@ -0,0 +1,149 @@
+import fs from 'fs';
+import path from 'path';
+import { NextFunction, Request, Response, Router } from 'express';
+import Ajv from 'ajv';
+
+import { Controller } from '../interfaces';
+
+import { documentValidationMiddleware } from '../middlewares/document-validation.middleware';
+
+import { ArchiverService } from '../services/archiver.service';
+import { GeneratorService } from '../services/generator.service';
+
+import { ProblemException } from '../exceptions/problem.exception';
+import { prepareParserConfig } from '../utils/parser';
+
+/**
+ * Controller which exposes the Generator functionality
+ */
+export class GenerateController implements Controller {
+  public basepath = '/generate';
+
+  private archiverService = new ArchiverService();
+  private generatorService = new GeneratorService();
+  private ajv: Ajv;
+
+  private async generate(req: Request, res: Response, next: NextFunction) {
+    try {
+      this.validateTemplateParameters(req);
+    } catch(err) {
+      return next(err);
+    }
+    
+    const zip = this.archiverService.createZip(res);
+
+    let tmpDir: string;
+    try {
+      tmpDir = this.archiverService.createTempDirectory();
+      const { asyncapi, template, parameters } = req.body;
+
+      await this.generatorService.generate(
+        req.parsedDocument,
+        template,
+        parameters,
+        tmpDir,
```

Comment:
I still think we should move into generating in memory, writing directly to the zip stream by using (If I'm correct) `append` archiver method: https://www.archiverjs.com/docs/archiver#append.

The whole process will be generated on the fly, and theoretically, since you piped the stream to the response object, data should be flushed on each write, isn't? Meaning you don't hold everything in memory but just a small buffer since everything is drained though the response (sent back to the client).

Reply:
Maybe I'm not understanding it correctly, but how to do that if generator uses under the hood filesystem api to "render" templates? For that I passed directory destination for output.